### ISSUE TITLE
Proposal: Add Visual Tree Helper Methods to GetParent/GetChild of a Type

### ISSUE THREAD

**Issue by robloo** [microsoft/microsoft-ui-xaml#1331]
# Proposal: Add Visual Tree Helper Methods to GetParent/GetChild of a Type

It seems there are lots of more advanced use cases where you need to get a visual tree parent or child of a specific type. I was reminded of this when reading the WPF article here:

https://docs.microsoft.com/en-us/dotnet/framework/wpf/data/how-to-find-datatemplate-generated-elements

I already have some UWP helper methods to implementing the "FindVisualChild" logic mentioned in this article but I think they should be added to UWP/WinUI and even WPF as well.

## Summary

There is no better summary than actual code which I would be happy to donate to the project under the MIT license. Of course that needs to get converted into C++ and would go in the VisualTreeHelper class along with the existing GetParent() method.

```csharp
        /// <summary>
        /// Finds the parent element of the specified type.
        /// </summary>
        /// <typeparam name="T">The type of the parent element to return.</typeparam>
        /// <param name="child">The child to find the element for.</param>
        /// <returns>The parent element or null.</returns>
        public static T GetVisualParent<T>(DependencyObject child) where T : DependencyObject
        {
            T parent = null;
            DependencyObject currParent;

            currParent = VisualTreeHelper.GetParent(child);
            while (currParent != null)
            {
               if (currParent is T)
               {
                  parent = (T)currParent;
                  break;
               }

               // find the next parent
               currParent = VisualTreeHelper.GetParent(currParent);
            }

            return (parent);
        }

        /// <summary>
        /// Finds the child element of the specified type and name.
        /// If no name is given, the first child of the specified type will be returned.
        /// </summary>
        /// <typeparam name="T">The type of the child element to return.</typeparam>
        /// <param name="parent">The parent to find the child within.</param>
        /// <param name="name">An optional name of the child element.</param>
        /// <returns>The child element or null.</returns>
        public static T GetVisualChild<T>(DependencyObject parent, 
                                          string name = null) where T : DependencyObject
        {
            T result;
            int count;
            DependencyObject currChild;
            FrameworkElement currChildElement;

            count = VisualTreeHelper.GetChildrenCount(parent);
            for (int i = 0; i < count; i++)
            {
                currChild = VisualTreeHelper.GetChild(parent, i);
                currChildElement = currChild as FrameworkElement;

                if (currChild is T)
                {
                    if (string.IsNullOrEmpty(name) == false)
                    {
                        // A name was given so the child must match it
                        if ((currChildElement != null) &&
                            (currChildElement.Name == name))
                        {
                            return ((T)currChild);
                        }
                    }
                    else
                    {
                        // No name was given but the type matches
                        return ((T)currChild);
                    }
                }

                // Recursively search children
                result = GetVisualChild<T>(currChild, name);
                if (result != null)
                {
                    return (result);
                }
            }

            return (null);
        }
```

## Rationale

Simplify development when you need to get an element from the visual tree.

**Comment by michael-hawker** [comment]
🦙 For reference, we have both [Visual](https://docs.microsoft.com/en-us/windows/communitytoolkit/extensions/visualtree) and [Logical](https://docs.microsoft.com/en-us/windows/communitytoolkit/extensions/logicaltree) tree extensions in the [Windows Community Toolkit](https://aka.ms/wct).

**Comment by robloo** [comment]
@michael-hawker I didn't realize that, it's good to know! This should still be part of the framework in my opinion. The Toolkit has more helpers so those are good API additions as well -- although it should be using the Parent/Child naming convention the same as VisualTreeHelper.

**Comment by michael-hawker** [comment]
Yeah @robloo the naming of the methods was done before my time on the toolkit.  But in the toolkit they're also extension methods to the elements themselves and not a helper.  They wanted to stay away from Parent/Child as that has connotations from the classes themselves (that's why I created the Logical tree ones which follow that naming instead).

In some ways the VisualTreeHelper naming is misleading; however, it is the VisualTreeHelper so the context of which tree is being navigated is inferred by usage.

**Comment by jevansaks** [comment]
The "framework" per se can't offer these because generic/templated stuff isn't well-defined in WinRT. Especially since there are options in the toolkit and searching the tree isn't ideal (we'd like to start locking down control trees _more_), I don't think we would add this.

**Comment by adrientetar** [comment]
> Especially since there are options in the toolkit

What do you mean by that?

**Comment by michael-hawker** [comment]
@adrientetar outside of some of the technical challenges @jevansaks mentioned, I believe they want to be able to restyle their controls as needed to be able to update and add new features.  They don't want to have official methods to make it easy for developers to reach-in to platform controls in order to manipulate them in the tree.

However, if you're a developer making your own control or working within your whole stack, these types of functions and abilities can be very handy if used properly (as there are performance implications to walking the tree).  So, for now, if you need them, you can use the ones in the toolkit.

If anything, maybe we should port them to a more consumable method in the toolkit for C++ developers, but someone would need to raise an issue on that in our repo, and we'd have to evaluate the best way to support that now and in the future.

**Comment by robloo** [comment]
@jevansaks, 

>  generic/templated stuff isn't well-defined in WinRT.
Good point! I didn't consider how complex that C# would be to implement in C++ let alone work through the language projection tooling. Generics are "well?" supported in C++ now though and perhaps this is something that should be better defined in the future. That's out of scope here though and I would be fine to close this issue due to technical reasons and the existence of the methods in the community toolkit.

> we'd like to start locking down control trees more
I understand the goal to lock down the Visual Tree which would allow, among other things, templates to be changed without having to worry about breaking apps. However, there are many cases where using the visual tree is useful for the end app itself and the structure of the tree is known from a DataTemplate.

I'll try to keep this example short, but I recently implemented an editor where there are multiple items in a list. Only when an item in the list is selected do editing controls appears (TextBox, ComboBox). This is easy enough with a custom DataTemplateSelector. However, it became clear that some of the editing controls are not light-weight (the list of items in ComboBox) and loading them at selection is not ideal. Therefore, as a performance optimization, the editing controls are loaded once and then re-parented in the UI where they are needed (since they are only used at the current selection once at a time). The issue was a reference is needed to the generated controls from the DataTemplate to swap in the editing controls at each selection. Without the VisualTreeHelper (and honestly a little hack waiting for the data template to actually generate controls) I would not have been able to do this it seems. UWP does not have a way to do this:

https://docs.microsoft.com/en-us/dotnet/framework/wpf/data/how-to-find-datatemplate-generated-elements
```
TextBlock myTextBlock = (TextBlock)myDataTemplate.FindName("textBlock", myContentPresenter);
```

Without the ability to navigate the Visual Tree I'm not sure how I would have got around this limitation. I'm sure I could think of more examples, but locking down the Visual Tree is not a good idea for more advanced apps where developers know the visual tree (they build the data templates).

**Comment by robloo** [comment]
I recently noticed that using the VisualTree is actually required in the NavigationView in some cases according to the documentation. Further evidence that locking down the VisualTree is generally problematic in some edge cases.

> To add access keys for the back arrow and settings buttons that are built-in to NavigationView, you have to use VisualTreeHelper to get a reference to those controls, then set the properties in your code.

https://docs.microsoft.com/en-us/uwp/api/microsoft.ui.xaml.controls.navigationview?view=winui-2.2

This is probably a bad implementation in the NavigationView though and I wonder if it should change.

**Comment by ranjeshj** [comment]
I agree that APIs for walking the tree would be useful and have had conversations with @Austin-Lamb on this before. As mentioned above apart from the template scenario, there are other cases where this could be useful. Also, walking the tree in the framework can be more performant than doing it in managed code.

**Comment by Austin-Lamb** [comment]
This would indeed be a very useful API, and by implementing it in the platform we can do it more efficiently as Ranjesh mentions.  This will need WinUI 3, so it's not something we can do soon, but it's a great suggestion.

**Comment by MikeHillberg** [comment]
I get the utility of having this and I've used it. But I also think it's generally a sign that something else is broken or missing. (It's a workaround much like private reflection in c#.)  

For example [in the WPF case](https://docs.microsoft.com/en-us/dotnet/framework/wpf/data/how-to-find-datatemplate-generated-elements) that @robloo pointed to, rather than having to do a `FindName("textBlock")` there should be a typed code-behind story such that you can just do `this.textBlock`.

**Comment by robloo** [comment]
@MikeHillberg, I wouldn't go as far as saying this indicates something is broken. Even the original WPF designers thought it was necessary in some edge cases. If you ever customize a control like DataGrid you have no choice but to build custom data templates and whether you do it XAML directly or code-behind XAML you sometimes need a way to get a reference to a generated control (especially when binding doesn't work like dynamically updating ComboBox items).

Of course if there was a typed way to do this as you suggest it would be great from a code safety standpoint. Even adding back the [FrameworkElementFactory](https://docs.microsoft.com/en-us/dotnet/api/system.windows.frameworkelementfactory?view=netframework-4.8) from WPF to build DataTemplates doesn't allow having references to generated controls. You would have to think of some new ideas here.

With a typed way of doing things it would probably be a lot of additional code for little benefit when other techniques (like above) work well enough. The cost/benefit equation probably doesn't make sense.

**Comment by MikeHillberg** [comment]
@robloo, sorry, I didn't mean something's broken/missing in your code, I meant we're missing a feature in Xaml. I'd just like these scenarios to be easier (and more strongly typed) than FindName lookups and Visual Tree walking.

**Comment by robloo** [comment]
@MikeHillberg No worries! I understood you were talking about the XAML platform itself. I still can't think of a good way to do what you suggest; even by adding new functionality to XAML itself. It would require some out-of-the-box new ideas to do I think.

My concerns about cost/benefit also apply to adding this to the platform.

**Comment by robloo** [comment]
Here is another need for walking the visual tree. This just helps solidify the need for improving functionality around this area and that it shouldn't be locked down in the future.

https://github.com/microsoft/microsoft-ui-xaml/issues/2620#issuecomment-641551233

**Comment by robloo** [comment]
And another use of this directly in WinUI where this request would work perfectly:

https://github.com/microsoft/microsoft-ui-xaml/blob/9c0115f57c21ba1eb7d9e01c5b400356c4d317c1/dev/ColorPicker/ColorPickerSlider.cpp#L197-L210

**Comment by robloo** [comment]
If this isn't on the roadmap at all I'm fine with closing it. There are alternatives in the toolkit and most just end up writing their own helper methods anyway.
